mlflow/store/sftp_artifact_repo.py

```diff
@@ -0,0 +1,89 @@
+import os
+
+from mlflow.entities.file_info import FileInfo
+from mlflow.store.artifact_repo import ArtifactRepository
+from mlflow.utils.file_utils import TempDir
+
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse
+
+
+class SFTPArtifactRepository(ArtifactRepository):
+    """Stores artifacts as files in a remote directory, via sftp."""
+
+    def __init__(self, artifact_uri, client=None):
+        self.uri = artifact_uri
+        parsed = urlparse(artifact_uri)
+        self.config = {
+            'host': parsed.hostname,
+            'port': 22 if parsed.port is None else parsed.port,
+            'username': parsed.username,
+            'password': parsed.password
+        }
+        self.path = parsed.path
+
+        if client:
+            self.sftp = client
+        else:
+            import pysftp
+            import paramiko
+
+            ssh_config = paramiko.SSHConfig()
+            user_config_file = os.path.expanduser("~/.ssh/config")
+            if os.path.exists(user_config_file):
+                with open(user_config_file) as f:
+                    ssh_config.parse(f)
+
+            user_config = ssh_config.lookup(self.config['host'])
+
+            if self.config['username'] is None and 'username' in user_config:
+                self.config['username'] = user_config['username']
+
+            if 'identityfile' in user_config:
+                self.config['private_key'] = user_config['identityfile'][0]
+
+            self.sftp = pysftp.Connection(**self.config)
+
+        super(SFTPArtifactRepository, self).__init__(artifact_uri)
+
+    def log_artifact(self, local_file, artifact_path=None):
+        artifact_dir = os.path.join(self.path, artifact_path) \
+            if artifact_path else self.path
+        self.sftp.makedirs(artifact_dir)
+        self.sftp.put(local_file, artifact_dir)
```

Comment:
Were you able to confirm if this overwrites `artifact_dir` with `local_file`? Or does it just put `local_file` into `artifact_dir` (the desired behavior)? Looking at the [pysftp](http://pysftp.readthedocs.io/en/release_0.2.9/pysftp.html#pysftp.Connection.put) docs it seems we might overwrite `artifact_dir`

Reply:
I was here wrong indeed. After adding a test, the error was confirmed.